### PR TITLE
Add SemiBin

### PR DESCRIPTION
**Pull request by psj1997**
## PR checklist

Closes #XXX 

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [ ] Remove all TODO statements.
- [ ] Emit the `versions.yml` file.
- [ ] Follow the naming conventions.
- [ ] Follow the parameters requirements.
- [ ] Follow the input/output options guidelines.
- [ ] Add a resource `label`
- [ ] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [ ] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [x] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

### CODE REVIEW

File: modules/nf-core/semibin/singleeasybin/main.nf
Comment (nvnieuwk) [inline modules/nf-core/semibin/singleeasybin/main.nf:41]: These should be specified using `ext.args2` since you have the main tool can also receive arguments. See here paragraph 1: https://nf-co.re/docs/contributing/modules#general

File: modules/nf-core/semibin/singleeasybin/main.nf
Comment (nvnieuwk) [inline modules/nf-core/semibin/singleeasybin/main.nf:44]:
```suggestion
```

File: modules/nf-core/semibin/singleeasybin/main.nf
Comment (nvnieuwk) [inline modules/nf-core/semibin/singleeasybin/main.nf:50]:
```suggestion
```

File: modules/nf-core/semibin/singleeasybin/main.nf
Comment (nvnieuwk) [inline modules/nf-core/semibin/singleeasybin/main.nf:49]:
```suggestion
    tuple val(meta), path("*.csv")                        , emit: csv
    tuple val(meta), path("*.h5")                         , emit: model
    tuple val(meta), path("output_prerecluster_bins/*.fa"), emit: output_fasta
    tuple val(meta), path("output_recluster_bins/*.fa")   , emit: recluster_fasta
    tuple val(meta), path("*.tsv")                        , emit: tsv
```

Comment (nvnieuwk) [reply modules/nf-core/semibin/singleeasybin/main.nf:49]: Some standardization :)

File: modules/nf-core/semibin/singleeasybin/main.nf
Comment (nvnieuwk) [inline modules/nf-core/semibin/singleeasybin/main.nf:67]:
```suggestion
```

File: modules/nf-core/semibin/singleeasybin/main.nf
Comment (nvnieuwk) [inline modules/nf-core/semibin/singleeasybin/main.nf:80]: So these should be in `ext.args2`

File: modules/nf-core/semibin/singleeasybin/meta.yml
Comment (nvnieuwk) [inline modules/nf-core/semibin/singleeasybin/meta.yml:2]:
```suggestion
```

File: modules/nf-core/semibin/singleeasybin/meta.yml
Comment (nvnieuwk) [inline modules/nf-core/semibin/singleeasybin/meta.yml:9]:
```suggestion
```

File: modules/nf-core/semibin/singleeasybin/meta.yml
Comment (nvnieuwk) [inline modules/nf-core/semibin/singleeasybin/meta.yml:17]:
```suggestion
```

File: modules/nf-core/semibin/singleeasybin/meta.yml
Comment (nvnieuwk) [inline modules/nf-core/semibin/singleeasybin/meta.yml:19]:
```suggestion
```

File: modules/nf-core/semibin/singleeasybin/meta.yml
Comment (nvnieuwk) [inline modules/nf-core/semibin/singleeasybin/meta.yml:26]:
```suggestion
```

File: modules/nf-core/semibin/singleeasybin/meta.yml
Comment (nvnieuwk) [inline modules/nf-core/semibin/singleeasybin/meta.yml:36]:
```suggestion
```

File: modules/nf-core/semibin/singleeasybin/meta.yml
Comment (nvnieuwk) [inline modules/nf-core/semibin/singleeasybin/meta.yml:38]:
```suggestion
```

File: modules/nf-core/semibin/singleeasybin/meta.yml
Comment (nvnieuwk) [inline modules/nf-core/semibin/singleeasybin/meta.yml:44]:
```suggestion
```

File: modules/nf-core/semibin/singleeasybin/meta.yml
Comment (nvnieuwk) [inline modules/nf-core/semibin/singleeasybin/meta.yml:74]:
```suggestion
  - csv:
      type: file
      description: generated files
      pattern: "*.csv"
  - h5:
      type: file
      description: trained model
      pattern: "*.h5"
  - output_prerecluster_bins:
      type: file
      description: output_prerecluster_bins
      pattern: "output_prerecluster_bins/*.fa"
  - output_recluster_bins:
      type: file
      description: output_recluster_bins
      pattern: "output_recluster_bins/*.fa" 
  - tsv:
      type: file
      description: information of bins
      pattern: "*.tsv"     
```

File: tests/modules/nf-core/semibin/singleeasybin/main.nf
Comment (nvnieuwk) [inline tests/modules/nf-core/semibin/singleeasybin/main.nf:30]: Isn't it possible to work with the BAM files already available in the test data repository? (less compute time for the tests is better :p)

File: tests/modules/nf-core/semibin/singleeasybin/test.yml
Comment (nvnieuwk) [inline tests/modules/nf-core/semibin/singleeasybin/test.yml:30]: Can you fill this in?

File: tests/modules/nf-core/semibin/singleeasybin/main.nf
Comment (psj1997) [reply tests/modules/nf-core/semibin/singleeasybin/main.nf:30]: I have tried, but it seems the bam files in the repo can not be used.

Comment (nvnieuwk) [reply tests/modules/nf-core/semibin/singleeasybin/main.nf:30]: Hmm strange since those BAMs were also aligned from those FastQ files. What was the error you got?

Comment (psj1997) [reply tests/modules/nf-core/semibin/singleeasybin/main.nf:30]: SemiBin needs the contig files (> 5 contigs) and corresponding bam file. But what I found is contig file (1 contig) and the bam file. If there are files that suits the needs in the repo? Thanks!

Comment (nvnieuwk) [reply tests/modules/nf-core/semibin/singleeasybin/main.nf:30]: Aaah no I see how you fixed it here! Yes this seems like the ideal solution indeed :)

Comment (nvnieuwk) [reply tests/modules/nf-core/semibin/singleeasybin/main.nf:30]: Thanks for clarifying :partying_face:

**Comment by nvnieuwk** [comment]
Checks are running!

**Comment by nvnieuwk** [comment]
@nf-core-bot fix linting

File: modules/nf-core/semibin/singleeasybin/main.nf
Comment (nvnieuwk) [inline modules/nf-core/semibin/singleeasybin/main.nf:8]: Are these correct? Your tests are failing because it can't find these containers. (possibly because of `%3A` in the name)

Comment (nvnieuwk) [reply modules/nf-core/semibin/singleeasybin/main.nf:8]: `docker: invalid reference format: repository name must be lowercase.`

Comment (psj1997) [reply modules/nf-core/semibin/singleeasybin/main.nf:8]: In the local testing, I have tested with singularity, it works.

Comment (psj1997) [reply modules/nf-core/semibin/singleeasybin/main.nf:8]: It needs to pass the testing in all three ways?

Comment (nvnieuwk) [reply modules/nf-core/semibin/singleeasybin/main.nf:8]: It's preffered that all tests pass, but the once for docker and singularity are really necessary to be sure it works almost everywhere

Comment (nvnieuwk) [inline modules/nf-core/semibin/singleeasybin/main.nf:8]:
```suggestion
        'https://depot.galaxyproject.org/singularity/semibin:1.4.0--pyh7cba7a3_0':
        'quay.io/biocontainers/semibin:1.4.0--pyh7cba7a3_0' }"
```

Maybe this will work?

Comment (psj1997) [reply modules/nf-core/semibin/singleeasybin/main.nf:8]: Thanks! 

For the conda, if there is a way to add another channel (conda-forge) here: conda "bioconda::semibin=1.4.0"?

Sincerely
Shaojun

Comment (psj1997) [reply modules/nf-core/semibin/singleeasybin/main.nf]: I think this maybe can work.

Comment (nvnieuwk) [reply modules/nf-core/semibin/singleeasybin/main.nf]: The problem with the conda should have been fixed now (by merging with the `master` branch) It was a known problem :)

File: modules/nf-core/semibin/singleeasybin/main.nf
Comment (nvnieuwk) [inline modules/nf-core/semibin/singleeasybin/main.nf:42]:
```suggestion
    "${task.process}": \$( SemiBin --version )
```

This should not be here :) (it's already in the `"${task.process}"`)

Comment (psj1997) [reply modules/nf-core/semibin/singleeasybin/main.nf]: Thank you! There seems another error of the `model.h5` file. I do not why but it can pass the testing locally.

Comment (nvnieuwk) [reply modules/nf-core/semibin/singleeasybin/main.nf]: I'll have a look if the tests ran :)

File: modules/nf-core/semibin/singleeasybin/main.nf
Comment (nvnieuwk) [inline modules/nf-core/semibin/singleeasybin/main.nf:43]:
```suggestion
    cat <<-END_VERSIONS > versions.yml
    "${task.process}": 
        SemiBin: \$( SemiBin --version )
    END_VERSIONS
```

I was wrong earlier, my bad. It should have been better indented I think, can you try this?

Comment (psj1997) [reply modules/nf-core/semibin/singleeasybin/main.nf]: OK. Thanks!

**Comment by psj1997** [comment]
If there is a better way for checking a binary fiile?

**Comment by nvnieuwk** [comment]
I'll approve if the tests pass :)

**Comment by psj1997** [comment]
Sorry. I do not know why the process stuck in the `Some checks haven’t completed yet`. Thanks!

**Comment by nvnieuwk** [comment]
This happens some time, usually pushing an empty/small commit fixes this

File: modules/nf-core/semibin/singleeasybin/main.nf
Comment (nvnieuwk) [inline modules/nf-core/semibin/singleeasybin/main.nf:42]:
```suggestion
    "${task.process}":
```

linting error :)

**Review by nvnieuwk — APPROVED** [review]
LGTM if that list fix is applied

**Comment by psj1997** [comment]
It seems that the issue of conda is about the channel (conda-forge).

**Comment by psj1997** [comment]
When using conda to install SemiBin, we need to specify conda-forge: `conda install -c conda-forge -c bioconda semibin` . If not, the install will fail. So I guess this is the issue.

But I do not know how to add another channel in the main.nf script. Do you have some suggestions?

Sincerely
Shaojun

**Comment by nvnieuwk** [comment]
`conda-forge` is already added in the CI command

**Comment by psj1997** [comment]
If conda-forge is in the channel, I have tested it locally and it runs successfully. I do not why for this issue.

**Comment by nvnieuwk** [comment]
This is the code used to set up conda in the CI tests:
```
      - name: Set up miniconda
        if: matrix.profile == 'conda'
        uses: conda-incubator/setup-miniconda@v2
        with:
          auto-update-conda: true
          channels: conda-forge,bioconda,defaults
          python-version: ${{ matrix.python-version }}
```

It seems to be a known problem with a fix in this PR: https://github.com/nf-core/modules/pull/2740#pullrequestreview-1251020662

So you can either wait for this or you can merge the code as the docker and singularity tests are fine :)

**Comment by psj1997** [comment]
Thanks. For me, I want to we can merge the code now.

**Comment by nvnieuwk** [comment]
Go ahead! The honor is yours! :partying_face:

**Comment by psj1997** [comment]
Sorry. I do not know what I can do now to help to merge the code?

**Comment by nvnieuwk** [comment]
Just press the Squash and merge button :)

**Comment by psj1997** [comment]
I do not see the merge button.

**Comment by nvnieuwk** [comment]
No problem, I pressed it now :) Congrats on your first module!

**Comment by psj1997** [comment]
Thanks for your time to help me to fix these issues!

**Pull request merged**
